src/docmanager/index.ts

```
@@ -0,0 +1,948 @@
+        parent.close();
+      });
+    });
+  }
+
+  /**
+   * Get whether the document manager has been disposed.
+   */
+  get isDisposed(): boolean {
+    return this._contentsManager === null;
+  }
+
+  /**
+   * Dispose of the resources held by the document manager.
+   */
+  dispose(): void {
+    if (this.isDisposed) {
+      return;
+    }
+    for (let modelName in this._modelFactories) {
+      this._modelFactories[modelName].factory.dispose();
+    }
+    this._modelFactories = null;
+    for (let widgetName in this._widgetFactories) {
+      this._widgetFactories[widgetName].factory.dispose();
+    }
+    this._widgetFactories = null;
+    for (let id in this._widgets) {
+      for (let widget of this._widgets[id]) {
+        widget.dispose();
+      }
+    }
+    this._widgets = null;
+    this._contentsManager = null;
+    this._sessionManager = null;
+    this._contextManager.dispose();
+    this._contextManager = null;
+  }
+
+  /**
+   * Register a widget factory with the document manager.
+   *
+   * @param factory - The factory instance.
+   *
+   * @param options - The options used to register the factory.
+   *
+   * @returns A disposable used to unregister the factory.
+   *
+   * #### Notes
+   * If a factory with the given `displayName` is already registered, the
+   * factory will be ignored and a warning will be printed to the console.
+   * If `'.*'` is given as a default extension, the factory will be registered
+   * as the global default.
+   * If a factory is already registered as a default for a given extension or
+   * as the global default, this factory will override the existing default.
+   */
+  registerWidgetFactory(factory: IWidgetFactory<Widget>, options: IWidgetFactoryOptions): IDisposable {
+    let name = options.displayName;
+    let exOpt = options as Private.IWidgetFactoryEx;
+    exOpt.factory = factory;
+    if (this._widgetFactories[name]) {
+      console.warn(`Duplicate registered factory ${name}`);
+      return;
```

This should throw an error, not just return. The thing it returns is not an IDisposable, after all.